src/core/libMeta.ml

```diff
@@ -0,0 +1,111 @@
+(** Functions to manipulate metavariables. *)
+
+open Term
+open Timed
+
+let meta_counter : int Stdlib.ref = Stdlib.ref (-1)
+
+(** [reset_counter ()] resets the counter used to produce meta keys. *)
+let reset_meta_counter () = Stdlib.(meta_counter := -1)
+
+(** [fresh p ?name a n] creates a fresh metavariable of type [a] and arity [n]
+   with the optional name [name], and adds it to [p]. *)
+let fresh : problem -> ?name:string -> term -> int -> meta =
+  fun p ?name a n ->
+  let m = {meta_key = Stdlib.(incr meta_counter; !meta_counter);
+           meta_name = name; meta_type = ref a; meta_arity = n;
+           meta_value = ref None } in
+  p.metas <- MetaSet.add m p.metas; m
```

Comment:
To avoid exposing the counter, the following can be written
```ocaml
let ((reset_meta_counter: unit -> unit), (fresh: problem -> ?name:string -> term -> int -> meta)) =
  let counter = Stdlib.ref (-1) in
  (fun () -> Stdlib.(counter := -1), fun p ?name a n -> ...)
```

Reply:
I prefer to add a libMeta.mli file, but this is not urgent and could be made in another PR.